src/scores/emerging/emerging_impl.py

```diff
@@ -0,0 +1,478 @@
+    fcst, obs, decision_weights, severity_dim, prob_threshold_dim, threshold_assignment, weights
+):
+    """
+    Checks that the risk matrix score inputs are suitable.
+    """
+    if severity_dim not in fcst.dims:
+        raise ValueError("`severity_dim` must be a dimension of `fcst`")
+    if severity_dim not in obs.dims:
+        raise ValueError("`severity_dim` must be a dimension of `obs`")
+    if severity_dim not in decision_weights.dims:
+        raise ValueError("`severity_dim` must be a dimension of `decision_weights`")
+    if weights is not None and severity_dim in weights.dims:
+        raise ValueError("`severity_dim` must not be a dimension of `weights`")
+    if prob_threshold_dim in fcst.dims:
+        raise ValueError("`prob_threshold_dim` must not be a dimension of `fcst`")
+    if prob_threshold_dim in obs.dims:
+        raise ValueError("`prob_threshold_dim` must not be a dimension of `obs`")
+    if prob_threshold_dim not in decision_weights.dims:
+        raise ValueError("`prob_threshold_dim` must be a dimension of `decision_weights`")
+    if weights is not None and prob_threshold_dim in weights.dims:
+        raise ValueError("`prob_threshold_dim` must not be a dimension of `weights`")
+    if len(decision_weights.dims) != 2:
+        raise ValueError("`decision_weights` must have exactly 2 dimensions: `severity_dim` and `prob_threshold_dim`")
+
+    if isinstance(fcst, xr.Dataset):
+        fcst = fcst.to_array()
+
+    if (fcst.max() > 1) or (fcst.min() < 0):
+        raise ValueError("values in `fcst` must lie in the closed interval `[0, 1]`")
+
+    if isinstance(obs, xr.Dataset):
+        obs = obs.to_array()
+
+    disallowed_obs_values = np.unique(obs.where(obs != 1).where(obs != 0).values)
+    disallowed_obs_values = disallowed_obs_values[~np.isnan(disallowed_obs_values)]
+    if len(disallowed_obs_values) > 0:
+        raise ValueError("values in `obs` can only be 0, 1 or nan")
+
+    if (decision_weights[prob_threshold_dim].min() <= 0) or (decision_weights[prob_threshold_dim].max() >= 1):
+        raise ValueError("`prob_threshold_dim` coordinates must be strictly between 0 and 1")
+
+    if not np.array_equal(np.sort(decision_weights[severity_dim].values), np.sort(fcst[severity_dim].values)):
+        raise ValueError("`severity_dim` coordinates do not match in `decision_weights` and `fcst`")
+    if not np.array_equal(np.sort(decision_weights[severity_dim].values), np.sort(obs[severity_dim].values)):
+        raise ValueError("`severity_dim` coordinates do not match in `decision_weights` and `obs`")
+
+    if threshold_assignment not in ["upper", "lower"]:
+        raise ValueError(""" `threshold_assignment` must be either \"upper\" or \"lower\" """)
+
+
+def _risk_matrix_score(
+    fcst: XarrayLike,
+    obs: XarrayLike,
+    decision_weights: xr.DataArray,
+    severity_dim: str,
+    prob_threshold_dim: str,
+    threshold_assignment: Optional[str] = "lower",
+) -> xr.DataArray:
```

Comment:
Add a "*" before threshold_assignment

Reply:
@nicholasloveday , I've made your suggested change.

@tennlee , I decided not to add a "*" here, because for `risk_matrix_score` you recommended to add a * after threshold_assignment. It seems odd to add one before threshold_assignment in  `_risk_matrix_score`

But will follow your direction on this if you disagree, as adding *s is new to me.

Reply:
Adding the "*" has an interesting effect. It requires that the keyword argument that follow must strictly be supplied as keyword arguments. In many ways, I wonder if the best practise would be to make every argument a keyword argument, and all following an "*", which would basically amount to supplying all arguments as a dictionary.

The alternative is that keyword arguments may be supplied as keywords, or, alternatively, using positional ordering. Consider a function `def fn(a, b, c=None, d=None): ...`

This could usually be called either with

`fn(1,2)
fn(1,2,4,3)
fn(1,2,4,d=3)
fn(1,2,(c=4, d=3)`

It's very easy to mix up the ordering of positionally inferred arguments. This is less error-prone for commonly patterned arguments (e.g. fcst, obs), and more so for arguments where the ordering is essentially arbitrary. Using the "*" would result in errors for the user when using positional ordering to supply keyword-arguments.

So, in short, the pattern in `scores` is to put a "*" before the first keyword argument to help cut down on programmer mistakes which would otherwise too-easily slip through.

It is also particularly important for the user-facing API and less so for internal code, but since it's generally better for coding accuracy all around, I think it should just be done everywhere.

I hope that helps explain.

Reply:
Thanks @tennlee , I've changed both functions so that there is a * before the kwargs